### ISSUE TITLE
Overriding templates

### ISSUE THREAD

**Issue by afbora** [lukasleitsch/kirby-blade#23]
Imagine I'm building a generic theme. But I need to overwrite some templates for a client. So I want it to look in multiple directories. And I want to put the templates I want to be overwritten under `/site/templates`.

### For example:

Default template for theme path: `/theme/templates/default.blade.php`
Override template path: `/site/templates/default.blade.php`

Is this possible or is it a feature request?

**Comment by fabianmichael** [comment]
I am currently working on a template and component registry in a feature branch, see https://github.com/lukasleitsch/kirby-blade/tree/feature/components-registry

The goal is to make it possible for themes to be a regular plugin. The new branch allows you to do the following:

```php

Kirby::plugin('my/theme', [
    'blade' => [
        'namespaces' => [
            'mynamespace => 'My\\Theme\\Components', // allows usage of <x-mynamespace::component-name />
        ],
        'templates' => __DIR__ . '/templates',
        'components' => [
          'demo' => Name\\Space\\SomeComponent::class, // Use class in arbitrary namespace as component
        ],
    ],
    'snippets => [
        'menu' => __DIR__ . '/snippets/menu.blade.php',
    ],
    'templates' => [
        'default' => __DIR__ . '/snippets/default.blade.php',
    ],
]);

```

As you can see, this gives you a ton of options for registering additional templates, snippets etc. The only thing I could not manage to implement are anonymous namespaced components. But apart from that, it already works very stable for a theme that I am currently working on. I decided to build it as a plugin, so updates can be handled through composer or a git submodule.

If you want to keep theme files within `/theme`, you could probably just write a companion plugin that registers the corrects paths.

**Comment by afbora** [comment]
Thank you for the information.

Then the overriding will be at plugin level, not possible at directory level. So in order to overwrite a template it will need to be registered via the plugin. Have I got it right?

**Comment by fabianmichael** [comment]
> Thank you for the information.
> 
> Then the overriding will be at plugin level, not possible at directory level. So in order to overwrite a template it will need to be registered via the plugin. Have I got it right?

If you set `blade.templates` in a plugin, that path will be searched by Blade when looking for templates and components first. You could also define single files by using the  `templates` and `snippets` keys instead, just like any other plugin in a non-Blade setup would. As soon, as the file also exists within the `site/templates` (or `site/snippets`) folder, that file is used instead. I tried to replicate how Kirby handles this out-of-the box: look for a template file within `site/templates` or if that does not exist: look for a plugin-provided template. I hope this answers your question?

**Comment by afbora** [comment]
````php
Kirby::plugin('my/theme', [
    'blade' => [
        'templates' => __DIR__ . '/templates',
    ],
]);
````

Ah, now I understand a little more. Considering I set it up like in your example, it will first look in the `site/templates` directory and then in the path specified in the plugin, right?

**Comment by afbora** [comment]
@fabianmichael When do you plan to release the `components-registry` feature?

**Comment by fabianmichael** [comment]
@afbora As soon as I find the time for updating the README … don’t know when that will be, sorry.

**Comment by fabianmichael** [comment]
Exactly, but currently this only works in the branch mentioned above and still needs documentation! :-)
